README.rst
=============================
django-fractions
=============================

.. image:: https://badge.fury.io/py/django-fractions.png
    :target: https://badge.fury.io/py/django-fractions

.. image:: https://github.com/jmichalicek/django-fractions/workflows/Python%20package/badge.svg
    :target: https://github.com/jmichalicek/django-fractions/actions?query=workflow%3A%22Python+package%22


Fraction display and form fields for Django

Documentation
-------------

The full documentation is at https://django-fractions.readthedocs.org.

Quickstart
----------

Install django-fractions::

    pip install django-fractions

Add ``djfractions`` to ``settings.INSTALLED_APPS``


Then use it in a project::

    import djfractions

In templates::

    {% load fractions %}
    {% display_fraction 1.25 %}

In Forms::

    from djfractions.forms import DecimalFractionField
    from django import forms

    class MyForm(forms.Form):
        a_fraction = DecimalFractionField()


Features
--------

* Template tag for displaying float and Decimal values as fractions including mixed numbers
* DecimalFractionField form field which handles input such as "1/4", "1 1/2", "1 and 1/2", and converts to a
  decimal.Decimal instance


TODO
-----

* Add unicode_fraction template tag to display the unicode fraction entity if available
* forms.FloatDecimalField to return a float rather than Decimal
* forms.SplitFractionWidget for having separate numerator and denominator form fields
* forms.SplitMixedFractionWidget for handling mixed number fractions with separate fields
* models.DecimalBackedFractionField() to store a Decimal value but return/accept it as a fraction
* models.FloatBackedFractionField() to store a Decimal value but return/accept it as a fraction


Cookiecutter Tools Used in Making This Package
----------------------------------------------

*  cookiecutter
*  cookiecutter-djangopackage


HISTORY.rst
.. :changelog:

History
-------

Current
+++++++++

5.0.0 (2023-01-08)
+++++++++
* Forgot about the 4.0.0 mess when I published 3.1.0, so starting fresh again.
  There are not really any backwards incompatible changes. See the 3.1.0 release notes.
* Fixing documentation and release history.

4.0.0 (2022-08-10)
+++++++++
* I'm dumb and goofed thing up and so this version exists on pypi

3.1.0 (2022-12-26)
++++++++++
* Add testing on python 3.11 by @jmichalicek in #30
* Fixed 'Polynomial regular expression used on uncontrolled data' codeQL issues which could be abused to impact performance.

3.0.0 (2022-08-10)
+++++++++
* Create codeql-analysis.yml by @jmichalicek in #21
* Add type checking by @jmichalicek in #22
* Update supported python versions to 3.7+ by @jmichalicek in #23
* fix: Django 4.0 deprecations by @GriceTurrble in #24
* Add testing of django 4.0 by @jmichalicek in #27
* Updated to support Django 4.1, drop old unsupported Django and Python versions by @jmichalicek in #29

2.0.0 (2020-09-13)
+++++++++
* Dropped support for Django 2.0 and lower and bump major version
* Dropped support for Python 3.4 and lower
* Updated DecimalFractionField to work on Django 2+ where from_db_field() does not take a context argument
  and where the field is expected to have a `context` attribute like DecimalField.
* Cleared out a bunch of Python 2 compatibility code such as how calls to super() are made, usage of six, and
  from __future__ imports.

1.1.0 (2017-06-04)
++++++++++++++++++
* add python 3.6 and django 1.11 to tox.ini - still a bit broken
* convert to matrix for environments in .travis.yml
  because tox only wants to test py3.6 when installed under 3.6
  but will not test 3.5 when running with python 3.6 as the base.
* Remove invalid ROOT_URLCONF from test django config
  There is no urls.py for djfractions, don't tell it to use one.  Older
  django versions were ok with this, but 1.11 is pickier about the correctness.
* add current changes to HISTORY.rst
* Adjust SILENCED_SYSTEM_CHECKS setting during tests
  Django 1.11 is stricter about system checks and will not even run
  the tests where there are some errors we specifically test for due
  to older django versions letting you make these mistakes.
* Added optional max_digits and decimal_places parameters to
  forms.DecimalFractionField so that returned Decimal objects have the
  desired max_digits and decimal_places when not directly tied to a
  models.DecimalField() on a ModelForm

1.0.0 (2016-12-31)
++++++++++++++++++
* Stop subclassing Django's DecimalField and duplicate small amounts of code
  as necessary for db backend compatibility.  Too many things need to be
  handled differently.  Main cause of major version bump.
* Update forms.FractionField to skip over max_digits and decimal_places kwargs which
  will get passed in by models.fields.DecimalFractionField
* Add models.fields.DecimalFractionField.formfield() so that a
  forms.FractionField will be used by default
* Fix quantity_to_decimal and quantity_to_fraction to strip leading and trailing
  spaces before pattern matching and converting to a decimal or fraction
* Allow for leading negative sign with forms.FractionField input values
* Fix is_fraction() to allow leading negative sign
* Add `max_digits` and `decimal_places` params to DecimalFractionField in test model
* Additional test cases for models.fields.DecimalFractionField


0.4.0 (2016-08-29)
++++++++++++++++++

* Added djfractions.models.DecimalFractionField which stores fractions.Fraction values as decimals in the dataase.
* Better usage of tox to test against different Python and Django versions
* Added testing against Django 1.10

0.3.2 (2015-08-28)
++++++++++++++++++

* Fixed boolean logic for when to coerce values to thirds in
  in forms.DecimalFractionField and get_fraction_parts()

0.3.1 (2015-08-12)
++++++++++++++++++

* HISTORY.rst typo fixes
* pypi release version fix

0.3.0 (2015-08-12)
++++++++++++++++++

* Added forms.FractionField which returns fractions.Fraction instances
* Refactoring of common code with new forms.FractionField
* Smarter checking for numeric types throughout the code
* forms.DecimalFractionField.to_python() handles fractions.Fraction values now
* Fixed bug handling negative numbers in quantity_to_decimal()
* Added min_value and max_value to forms.DecimalFractionField
* Made coerce_thirds, limit_denominator, and use_mixed_numbers params to DecimalFractionField
  proper named parameters and not just kwargs.

0.2.1 (2015-08-06)
++++++++++++++++++

* Fixed typo in usage docs

0.2.0 (2015-08-06)
++++++++++++++++++

* display_fraction template tag output is templated so that its formatting can be changed by users
* Added new display_improper_fraction template tag to simplify the common case of wanting to only use
  improper fractions with no whole numbers
* Added unicode_entity to template context for display_fraction and display_improper_fraction so that
  the html entity for common fractions may be used rather than <sup> and <sub> tags
* Refactored lots of code out into smaller, reusable functions
* Added a bunch of test cases

0.1.0 (2015-08-01)
++++++++++++++++++

* First release on PyPI.


djfractions/models/fields.py
import decimal
import fractions
import logging
from typing import Any, List, Optional, Tuple, Union

from django.core import checks
from django.core.checks.messages import CheckMessage
from django.db.models import Field
from django.utils.functional import cached_property
from django.utils.translation import gettext_lazy as _

from djfractions import coerce_to_thirds
from djfractions import forms as fraction_forms

logger = logging.getLogger(__name__)


class DecimalFractionField(Field):
    """
    Field which stores values as a Decimal value, but uses
    :class:`fractions.Fraction` for its value
    """

    empty_strings_allowed = False
    # ignoring type checking here because mypy and gettextlazy and django-stubs are fighting
    # possibly only on python 3.7 or django 3.2
    default_error_messages = {
        "invalid": _("'%(value)s' value must be a fraction number."),  # type: ignore
    }
    description = _("Fraction number stored in the database as a Decimal")

    def __init__(
        self,
        verbose_name: Optional[str] = None,
        name: Optional[str] = None,
        max_digits: Optional[int] = None,
        decimal_places: Optional[int] = None,
        limit_denominator: Optional[int] = None,
        coerce_thirds: bool = True,
        **kwargs
    ):
        self.limit_denominator = limit_denominator
        self.coerce_thirds = coerce_thirds

        # for decimal stuff
        self.max_digits, self.decimal_places = max_digits, decimal_places
        super().__init__(verbose_name=verbose_name, name=name, **kwargs)

    @cached_property
    def context(self) -> decimal.Context:
        # TODO: understand what this is and why it now exists and when and why it got added
        return decimal.Context(prec=self.max_digits)

    def check(self, **kwargs) -> List[CheckMessage]:
        errors = super().check(**kwargs)

        digits_errors = self._check_decimal_places()
        digits_errors.extend(self._check_max_digits())
        if not digits_errors:
            errors.extend(self._check_decimal_places_and_max_digits(**kwargs))
        else:
            errors.extend(digits_errors)
        return errors

    def _check_decimal_places(self) -> List[checks.Error]:
        try:
            # ignoring type to make mypy happy, the exception handling deals with this being incorrect
            decimal_places = int(self.decimal_places)  # type: ignore
            if decimal_places < 0:
                raise ValueError()
        except TypeError:
            return [
                checks.Error(
                    "DecimalFractionFields must define a 'decimal_places' attribute.",
                    obj=self,
                    id="fields.E130",
                )
            ]
        except ValueError:
            return [
                checks.Error(
                    "'decimal_places' must be a non-negative integer.",
                    obj=self,
                    id="fields.E131",
                )
            ]
        else:
            return []

    def _check_max_digits(self) -> List[checks.Error]:
        try:
            # ignoring type to make mypy happy, the exception handling deals with this being incorrect
            max_digits = int(self.max_digits)  # type: ignore
            if max_digits <= 0:
                raise ValueError()
        except TypeError:
            return [
                checks.Error(
                    "DecimalFractionFields must define a 'max_digits' attribute.",
                    obj=self,
                    id="fields.E132",
                )
            ]
        except ValueError:
            return [
                checks.Error(
                    "'max_digits' must be a positive integer.",
                    obj=self,
                    id="fields.E133",
                )
            ]
        else:
            return []

    def _check_decimal_places_and_max_digits(self, **kwargs) -> List[checks.Error]:
        # ignoring type to make mypy happy, the exception handling deals with this being incorrect
        if int(self.decimal_places) > int(self.max_digits):  # type: ignore
            return [
                checks.Error(
                    "'max_digits' must be greater or equal to 'decimal_places'.",
                    obj=self,
                    id="fields.E134",
                )
            ]
        return []

    def from_db_value(self, value: Any, expression: Any, connection: Any, *args, **kwargs) -> fractions.Fraction:
        # uses *args and **kwargs to handle the `context` param which django 1.11 passes in but 2.x+ do not.
        # Not sure if I even really need this anymore.
        return self.to_python(value)

    def get_db_prep_save(self, value: Any, connection):
        # return connection.ops.adapt_decimalfield_value(self.to_python(value), self.max_digits, self.decimal_places)
        # for django 1.9 the following will need used.
        if hasattr(connection.ops, "adapt_decimalfield_value"):
            return connection.ops.adapt_decimalfield_value(
                self.get_prep_value(value),
                self.max_digits,
                self.decimal_places,
            )
        else:
            return connection.ops.value_to_db_decimal(self.get_prep_value(value), self.max_digits, self.decimal_places)

    def to_python(self, value: Union[fractions.Fraction, decimal.Decimal, float, int, str]) -> fractions.Fraction:
        if value is None:
            return value

        # probably need similar error handling to
        # https://github.com/django/django/blob/stable/1.8.x/django/db/models/fields/__init__.py#L1598
        return self.to_fraction(value)

    def get_prep_value(
        self, value: Union[fractions.Fraction, decimal.Decimal, float, int, str, None]
    ) -> Union[decimal.Decimal, None]:
        # not super clear, docs sound like this must be overridden and is the
        # reverse of to_python, but
        # django.db.models.fields.DecimalField just calls self.to_python() here
        # TODO:
        # see https://docs.djangoproject.com/en/1.8/howto/custom-model-fields/#converting-python-objects-to-query-values
        # for usage.
        if value is None:
            return value

        if isinstance(value, fractions.Fraction):
            value = float(value)

        return decimal.Decimal(value)

    def to_fraction(self, value: Union[fractions.Fraction, decimal.Decimal, float, int, str]) -> fractions.Fraction:
        fraction_value = fractions.Fraction(value)

        if self.limit_denominator:
            fraction_value = fraction_value.limit_denominator(self.limit_denominator)

        if self.coerce_thirds and (not self.limit_denominator or self.limit_denominator > 3):
            fraction_value = coerce_to_thirds(fraction_value)

        return fraction_value

    def deconstruct(self) -> Tuple[str, str, list, dict]:
        name, path, args, kwargs = super().deconstruct()
        kwargs["limit_denominator"] = self.limit_denominator
        kwargs["coerce_thirds"] = self.coerce_thirds

        # added this
        # copied from decimal field
        if self.max_digits is not None:
            kwargs["max_digits"] = self.max_digits
        if self.decimal_places is not None:
            kwargs["decimal_places"] = self.decimal_places

        return name, path, args, kwargs

    def formfield(
        self,
        form_class: Optional[Any] = fraction_forms.FractionField,
        choices_form_class: Optional[Any] = None,
        **kwargs: Any
    ) -> Any:
        return super().formfield(form_class=form_class, choices_form_class=choices_form_class, **kwargs)

    def get_internal_type(self) -> str:
        # returning DecimalField, since we use the same backing column type as that
        # and this is safer than overriding db_type() and db_check() since that would
        # require maintaining the mapping for every db adapter.
        return "DecimalField"

    def _format(self, value: Union[decimal.Decimal, str, None]) -> Union[str, None]:
        if isinstance(value, str):
            return value
        else:
            return self.format_number(value)

    def format_number(self, value: Union[decimal.Decimal, None]) -> Union[str, None]:
        """
        Formats a number into a string with the requisite number of digits and
        decimal places.
        """
        # Method moved to django.db.backends.utils.
        #
        # It is preserved because it is used by the oracle backend
        # (django.db.backends.oracle.query), and also for
        # backwards-compatibility with any external code which may have used
        # this method.
        from django.db.backends import utils

        return utils.format_number(value, self.max_digits, self.decimal_places)
